### PR TITLE
add core concepts of parameter server part in ElasticDL

### PR DESCRIPTION
**Pull request by QiJune**
This PR aims to clarify the core concepts of parameter server part in ElasticDL. The doc will be updated continually. Any comments are welcomed.

### CODE REVIEW

File: docs/designs/concepts.md
Comment (skydoorkai) [inline docs/designs/concepts.md]: There two kinds  -> There are two kinds ...

File: docs/designs/concepts.md
Comment (skydoorkai) [inline docs/designs/concepts.md]: indluce -> include
a extra -> an extra

File: docs/designs/concepts.md
Comment (skydoorkai) [inline docs/designs/concepts.md]: In ElasticDL, only a large embedding layer is represented by an embedding table.  For small embedding layer, a dense parameter is used, but its gradient will have indices.

File: docs/designs/concepts.md
Comment (QiJune) [inline docs/designs/concepts.md]: Once step updating increase the version one time. The training of a Bert-large model need 1M steps udpating. So int32 is enough here.

File: docs/designs/concepts.md
Comment (QiJune) [inline docs/designs/concepts.md]: The key of Embeddings is int64. The embeddings maps from an item id to an embedding vector.

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (brightcoder01) [inline docs/designs/concepts.md]: The components of dense gradient -> What's the meaning of component?

File: docs/designs/concepts.md
Comment (brightcoder01) [inline docs/designs/concepts.md]: For better query and increasing embedding vectors => In order to query and insert embedding vectors efficiently

File: docs/designs/concepts.md
Comment (brightcoder01) [inline docs/designs/concepts.md]: Add comments for each property here?

File: docs/designs/concepts.md
Comment (brightcoder01) [reply docs/designs/concepts.md]: send back -> sent back

File: docs/designs/concepts.md
Comment (brightcoder01) [inline docs/designs/concepts.md]: This sentence is a little confusing

File: docs/designs/concepts.md
Comment (brightcoder01) [inline docs/designs/concepts.md]: int64 is missing?

File: docs/designs/concepts.md
Comment (brightcoder01) [inline docs/designs/concepts.md]: It has indices, which is actually an embedding table gradient. => It's the gradients for the embedding vectors corresponding to the indices?

File: docs/designs/concepts.md
Comment (lwpyr) [reply docs/designs/concepts.md]: Could fields or properties be more suitable?

Comment (QiJune) [reply docs/designs/concepts.md]: It means data structure. The data structure of dense gradient is the same with dense param.

File: docs/designs/concepts.md
Comment (lwpyr) [reply docs/designs/concepts.md]: This is a little bit confusing, is `For small embedding layer, a dense parameter is used` locally?

File: docs/designs/concepts.md
Comment (lwpyr) [reply docs/designs/concepts.md]: Referenced from Wikipedia
`Machine learning algorithms build a mathematical model based on sample data, known as "training data", in order to make predictions or decisions without being explicitly programmed to perform the task.`

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Remove it :)

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Rewrite the sentence.

File: docs/designs/concepts.md
Comment (skydoorkai) [inline docs/designs/concepts.md]: so `The data structure of dense gradient is the same with dense parameter` is not true.
Need to add the exception there.

`and so is the embedding table gradient`. Make it clear, not also the same with dense parameter, but the same with embedding table.

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: In-memory Representaiton => Tensor

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: TensorDType => TensorElementType

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: Do we need stride?

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: What are Indices?

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: DenseParam => DenseParameter

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: Don't use the alias, which usually implies that we are not clear with the naming.

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: Embedding => Tensor

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: I don't see that we need type Buffer, do we?  If not necessary, let us follow Occam's Razor.

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: I would suggest that we use protobuf instead of Go code in the design doc because protobuf is designed to be an IDL.

File: docs/designs/concepts.md
Comment (wangkuiyi) [inline docs/designs/concepts.md]: I agree to use map here in Model.  It seems that we don't need type DenseParameter={Tensor+name} any more.  Instead, we can have

```protobuf
syntax = "proto3";

message Model {
    map<string, Tensor> DenseParameters = 1,
}
```

**Review by wangkuiyi — CHANGES_REQUESTED** [review]

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: @wangkuiyi Here, I want to clarify the reason why we still need some extra in-memory data structure in Go, especially for embedding table. There are two reasons:

1.  Embedding table is usually a large dense tensor. In forward pass of embedding layer, worker will not pull the full table, but sends an array of item IDs to PS. Then the PS looks up the embeddig table, and sends corresponding embedding vectors back to worker. The worker will get a subset of the whole table. In the backward pass of embeddig layer, the gradients are also associated with IDs. Only IDs which are used in forward pass make contribution to the gradients.

TensorFlow introduce a data structure called [IndexedSlices](https://www.tensorflow.org/api_docs/python/tf/IndexedSlices) to represent the subset of a big dense tensor.
 The IndexedSlices has two filed, Tensor and Indices. Indices is a 1-dim array.

```
dense[indices[i],:] = tensor[i,:] 
```

2.  In our design, the embedding table is initialized lazily in PS. And the size of embedding table is also unknown. So we could not create a big dense tensor simplely. When looking IDs from workers, if the ID is not found in the table, the corresponding embedding vector will be created and initialized.  In order to query and insert embedding vectors efficiently in PS, we use a map data structure to represent EmbeddingTable.

In PS side, the embedding table parameter is a map data structure. In worker side, we pull the subset of the embedding table from PS, and push the gradients associated with IDs to PS.

We define a `IndexedSlices` message for the embedding table related RPC between worker and PS.

Following is the design:

```proto
message Tensor {
  repeated int64 dim = 1;
  bytes content = 2;
  TensorElementType dtype = 3;
}

message IndexedSlices {
  Tensor tensor = 1;
  repeated int64 indices = 2;
}

message Model {
  int32 version = 1; // model updated times
  map<string, Tensor> dense_parameters = 2;
  map<string, IndexedSlices> indexed_slices = 3;
}
```

For PS side, we propose following in-memory data structure in Go:

```go
import "elasticdl.org/elasticdl/pkg/proto"

type EmbeddingTable struct {
  Name             string
  Dim              int64
  Initializer      string
  EmbeddingVectors map[int64]*proto.Tensor
}

type Model struct {
  Version           int32
  InitStatus        bool
  DenseParameters   map[string]*proto.Tensor
  EmbeddingTables   map[string]*EmbeddingTable
}
```

File: docs/designs/concepts.md
Comment (skydoorkai) [inline docs/designs/concepts.md]: embedding table = tensor + index + name
So this `index` is not ID but address? 
What is the relationship with IndexedSlices? Is `indices` in `IndexedSlices` is this `index`?
```
message IndexedSlices {
  Tensor tensor = 1;
  repeated int64 indices = 2;
}
```

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Please refer to the comment [here](https://github.com/sql-machine-learning/elasticdl/pull/1715/files#r376240095)

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Please refer to the comment [here](https://github.com/sql-machine-learning/elasticdl/pull/1715/files#r376240095).

After we reach a consensus, I will update the doc.

File: docs/designs/concepts.md
Comment (wangkuiyi) [reply docs/designs/concepts.md]: After the phone call, I got the point.

On the parameter server, we'd prefer to maintain each embedding table as a map from ID to embedding vectors.  With such a data structure, it is efficient to allocate memory for new embedding vectors.  On the contrary, we'd concatenate embedding vectors into protobuf messages for parameter pulling and gradient pushing.  We cannot use concatenated embedding vectors as the in-memory data structure on the PS, because allocating new embedding vectors involves resize the space of concatenated embedding vectors.

Comment (wangkuiyi) [reply docs/designs/concepts.md]:
```protobuf
message IndexedSlices {
  Tensor concat_embedding_vecs = 1;
  repeated uint64 ids = 2;
}
```

**Review by wangkuiyi — APPROVED** [review]

**Comment by wangkuiyi** [comment]
I am approving this PR to dismiss my previous request for change. Please go on following the comments before merging. Thanks!

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done, remove it

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done, remove it

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Remove it, since we use `map<string, Tensor>` as DenseParameter

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Done

File: docs/designs/concepts.md
Comment (QiJune) [reply docs/designs/concepts.md]: Currently, we do not need stride.

The tensor is used for parameter pulling and gradient pushing. The paramter and gradient are all contious data.

**Review by terrytangyuan — APPROVED** [review]

**Pull request merged**
